### ISSUE TITLE
Write up useless / ambiguous flow control in lambdas

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#14591]
## Is your feature request related to a problem? Please describe.

```ruby
n = lambda do
  print 'n'
  next
end

b = lambda do
  print 'b'
  break
end

2.times do
  n.call
  print it
end
print ' '

2.times do
  b.call
  print it
end
puts ' DONE'
```

An unseasoned programmer may think `next` and `break` will work in the calling scope, especially since there's no flow logic in the lambda, so they wouldn't make sense there in the first place.

However, it in fact works in the lambda, and furthermore has no function there since it's at the end of the scope.

## Describe the solution you'd like

A cop should raise at least one eyebrow at the peculiarity.

## Describe alternatives you've considered

This possibly even raises the question if the Ruby interpreter should raise an exception for meaningless flow control. (Unless someone knows an official statement from maintainers that flow control keywords are graceful by design)

## Additional context

```
ruby      2025-08-01 30a20bc16
rubocop   2025-10-04 a08091813
System Version: macOS 15.6 (24G84)
Model Identifier: MacBookAir10,1
```

**Comment by forthrin** [comment]
@dvandersluis: Do you feel this one should be addressed?

**Comment by dvandersluis** [comment]
Yes actually, I was taking a look at this last week but haven't had a chance yet.

**Comment by forthrin** [comment]
@dvandersluis: Passing the six month marker. Happening, or not worth the effort?

**Comment by forthrin** [comment]
@dvandersluis: My good friend LLM would like to know how you feel about the following:

---

The proposal identifies a real readability issue, but the cop should target **provably redundant terminal flow control**, not unusual `next` or `break` generally.

### Recommended cop

Add a narrowly scoped cop such as:

```text
Lint/RedundantFlowControl
```

or extend an existing redundant-control-flow cop if one already covers the construct.

Report only when:

- `next` or `break` is the final executable statement in a lambda/block.
- It has no argument.
- It is not inside a `rescue`, `ensure`, or other control-flow-sensitive construct.
- Removing it preserves the block’s result and behavior.
- The construct is statically known to be a lambda or ordinary block.

Example:

```ruby
n = lambda do
  print "n"
  next
end
# Lint/RedundantFlowControl: Remove redundant `next` at the end of the lambda
```

Corrected:

```ruby
n = lambda do
  print "n"
end
```

### Treat `break` more cautiously

`break` is semantically more surprising than `next`, particularly in non-lambda blocks. It should initially be reported only for a terminal `break` inside a lambda:

```ruby
b = lambda do
  print "b"
  break
end
# Lint/RedundantFlowControl: Remove redundant `break` at the end of the lambda
```

Avoid flagging:

```ruby
items.each do |item|
  break if finished?(item)
end
```

Even a terminal unconditional `break` in a regular block may have non-obvious nonlocal control-flow behavior and should be excluded initially.

### Implementation strategy

1. Identify `next` and `break` nodes.
2. Determine their enclosing block or lambda.
3. Confirm they are the final effective statement.
4. Exclude nodes with arguments.
5. Exclude `ensure`, `rescue`, modifiers, and ambiguous dynamic contexts.
6. Provide autocorrection only for the simplest terminal cases.
7. Add tests covering nested blocks, lambdas, `Proc`s, `ensure`, conditionals, and loops.

Do not change the Ruby interpreter. The behavior is intentional language semantics, and a linter is better suited to flagging code that is valid but misleading.

The cost-benefit is favorable if the cop remains narrow. Detecting all “meaningless” flow control would be expensive and prone to false positives; detecting explicit terminal `next` and lambda `break` is comparatively cheap and actionable.
